### ISSUE TITLE
Error on SaveChanges() related to #132

### ISSUE THREAD

**Issue by marcioduarte89** [thepirat000/Audit.NET#273]
Hi,

i'm having the exact same issue as reported in [132](https://github.com/thepirat000/Audit.NET/issues/132).

We have multiple classes inheriting from Mapping (ClientMapping, DivisionMapping, ProductMapping), although we only have a Mappings table with a EntityType discriminator. The issue happens when trying to create and save an instance of a ClientMapping through our db context. 

These are our classes:

```csharp
public abstract class Mapping 
{
      public int Id { get; private set; }

      public string PropertyOne { get; private set; }

      public string PropertyTwo { get; private set; }

      // Other properties
}

```

   ```csharp
public class MappingConfiguration : EntityTypeConfiguration<Mapping> 
{
      public MappingConfiguration() 
      {
         ToTable("Mapping");

         Map<ClientMapping>(m => m.Requires("EntityType").HasValue(nameof(Client))); // EntityType is the discrimitator
         Map<DivisionMapping>(m => m.Requires("EntityType").HasValue(nameof(Division)));
         Map<ProductMapping>(m => m.Requires("EntityType").HasValue(nameof(Product)));

         Property(m => m.PropertyOne)
            .IsRequired();

         Property(m => m.PropertyTwo)
            .IsRequired()
            .HasMaxLength(50);
      }
   }
```

```csharp
public class ClientMapping : Mapping 
{
 // No new properties here atm

}
```

```csharp
public class DivisionMapping : Mapping 
{
 // No new properties here atm

}
```

```csharp
public class ProductMapping : Mapping 
{
 // No new properties here atm

}
```

I've done some investigation and the issue happens in EntityKeyHelper.cs on GetMappingFragment in https://github.com/thepirat000/Audit.NET/blob/74f3ee292d5b23dff9d0bc00593c9cccefb7b2b5/src/Audit.EntityFramework/EntityKeyHelper.cs#L112

But the issue starts when attempting (for ClientMapping):

```csharp
var columnName = mappingFragment
            .PropertyMappings
            .OfType<ScalarPropertyMapping>()
                  .SingleOrDefault(m => m.Property.Name == propertyName)?
                .Column
                .Name;
```

in EntityKeyHelper.GetColumnName

columnName is null, only the Id is returned from the MappingFragment. It then tries to get Column name for the Base Entity type (Mapping). When getting inside GetMappingFragment with the base type, the mapping.EntityTypeMappings collection is still holding 4 instances (Mapping, ClientMapping, DivisionMapping and ProductMapping) and all of them have at least one fragment.. An exception with "Sequence contains more than one matching element" is then raised.

Please let me know if you need more information.

Thank you,
Marcio

**Comment by thepirat000** [comment]
Please share your DbContext code, and any other mapping code you have

**Comment by marcioduarte89** [comment]
Thanks for replying.

Because of confidentiality, I obviously can't share all the details on the DbContext, but i can provide all relevant ones so you can (hopefully) reproduce the issue. I've been testing the Mapping classes and table in isolation without the interference from any other Entities:

Audit.NET configuration:

```csharp
var isAuditActive = ConfigurationManager.AppSettings["Audit.IsActive"];

 if (isAuditActive != null) 
 {
	bool.TryParse(isAuditActive, out var isActive);
	Configuration.AuditDisabled = !isActive;
 }

 Audit.EntityFramework.Configuration.Setup()
	.ForContext<ProjectContext>(c => c
	   .IncludeEntityObjects()
	   .IncludeIndependantAssociations())
	.UseOptOut()
	.IgnoreAny(t =>  t.Name.EndsWith("Audit"));

 Configuration.Setup()
	.UseEntityFramework(_ => _
	   .AuditTypeExplicitMapper(m => m
		  .Map<Entity, Audit>(AuditServices.EntityAudit.BuildAuditEntry) // example mapping
	   .AuditEntityAction<Core.Audit.Audit>(AuditServices.Audit.Build))
	   .IgnoreMatchedProperties());
```


```csharp
public class ProjectAuditContext : AuditDbContext 
{
      public ProjectAuditContext(string nameOrConnectionString) : base(nameOrConnectionString) 
      {
      }

      public override Task<int> SaveChangesAsync() 
      {
         return SaveChangesAsync(CancellationToken.None);
      }

      public override async Task<int> SaveChangesAsync(CancellationToken cancellationToken) 
      {
          return await base.SaveChangesAsync(cancellationToken));
      }
}
```

```csharp
public class ProjectContext : ProjectAuditContext, IProjectContext 
{
	public ProjectContext(string nameOrConnectionString) : base(nameOrConnectionString) 
	{
	}

	// Other Sets here..
	public DbSet<Audit> Audit { get; set; }
	public DbSet<Mapping> Mapping { get; set; }


	protected override void OnModelCreating(DbModelBuilder modelBuilder) 
	{
	 modelBuilder.Configurations.AddFromAssembly(Assembly.GetExecutingAssembly());
	}

	public override Task<int> SaveChangesAsync() 
	{
	 return SaveChangesAsync(CancellationToken.None);
	}

	public override async Task<int> SaveChangesAsync(CancellationToken cancellationToken) 
	{
	 return await base.SaveChangesAsync(cancellationToken);
	}
}
```

We obviously need to generate a audit entry based on ClientMapping, DivisionMapping or ProductMapping.

Thanks for your time and help,
Marcio

**Comment by thepirat000** [comment]
Also related to #90 

I'm working on a fix

**Comment by thepirat000** [comment]
This is fixed on version 15.0.4, please upgrade your references and re-test

Fix was on `EntityKeyHelper` adding an extra validation [here](https://github.com/thepirat000/Audit.NET/commit/40e838ac51e2716ef1ed5fd3a89b9061a345f693#diff-432388f8798e12549a5ad355280de04bR110)
